core/integration/kafka/plugin.go

```diff
@@ -0,0 +1,311 @@
+	for _, v := range p.envsInRunning {
+		array = append(array, v)
+	}
+	return array
+}
+
+func (p *Plugin) ProduceMessage(message []byte, topic string, envId string) {
+	log.WithField("partition", envId).
+		Debug("Producing a new kafka message on topic ", topic)
+
+	deliveryChannel := make(chan kafka.Event)
+	err := p.kafkaProducer.Produce(&kafka.Message{
+		TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
+		Value:          message,
+	}, deliveryChannel)
+
+	if err != nil {
+		log.WithField("partition", envId).
+			Error("Kafka message delivery failed: ", err.Error())
+	}
+
+	e := <-deliveryChannel
+	m, ok := e.(*kafka.Message)
+
+	if !ok {
+		log.WithField("partition", envId).Error("Could not read Kafka message delivery status")
+	} else if m.TopicPartition.Error != nil {
+		log.WithField("partition", envId).Error("Kafka message delivery failed: ", m.TopicPartition.Error)
+	} else {
+		log.WithField("partition", envId).Debugf("Kafka message delivered to topic %s [%d] at offset %v\n",
+			*m.TopicPartition.Topic, m.TopicPartition.Partition, m.TopicPartition.Offset)
+	}
+}
+
+func (p *Plugin) CreateUpdateCallback(varStack map[string]string) func() string {
+	return func() (out string) {
+		// Retrieve and update the env info
+		timestamp := uint64(time.Now().UnixMilli())
+		envInfo := p.NewEnvStateObject(varStack)
+		p.UpdateRunningEnvList(envInfo)
+
+		// Prepare and send new state notification
+		log.WithField("partition", envInfo.EnvironmentId).
+			Debug("Advertising the environment state (" + envInfo.State + ") and active runs list to Kafka")
+		newStateNotification := &kafkapb.NewStateNotification{
+			EnvInfo:   envInfo,
+			Timestamp: timestamp,
+		}
+		nsnData, err := proto.Marshal(newStateNotification)
+		if err != nil {
+			log.WithField("partition", envInfo.EnvironmentId).
+				Error("Could not marshall a new state notification: ", err)
+		}
+		p.ProduceMessage(nsnData, p.FSMTransitionTopic(envInfo.State), envInfo.EnvironmentId)
+
+		// Prepare and send active run list
+		activeRunsList := &kafkapb.ActiveRunsList{
+			ActiveRuns: p.GetRunningEnvList(),
+			Timestamp:  timestamp,
+		}
+		arlData, err := proto.Marshal(activeRunsList)
+		if err != nil {
+			log.WithField("partition", envInfo.EnvironmentId).
+				Error("Could not marshall an active runs list: ", err)
+		}
+		p.ProduceMessage(arlData, p.ActiveRunsListTopic(), envInfo.EnvironmentId)
+		return
+	}
+}
+
+func (p *Plugin) CallStack(data interface{}) (stack map[string]interface{}) {
+	call, ok := data.(*callable.Call)
+	if !ok {
+		return
+	}
+	varStack := call.VarStack
+
+	stack = make(map[string]interface{})
+	stack["Standby"] = p.CreateUpdateCallback(varStack)
+	stack["Deployed"] = p.CreateUpdateCallback(varStack)
+	stack["Configured"] = p.CreateUpdateCallback(varStack)
+	stack["Running"] = p.CreateUpdateCallback(varStack)
+	stack["Error"] = p.CreateUpdateCallback(varStack)
+	stack["Destroy"] = p.CreateUpdateCallback(varStack)
```

Comment:
I see that the callback builder only depends on the `varStack`. In this context, what's the impediment to just having a single key-value in the `stack` of functions, e.g. `stack["PublishUpdate"] = p.CreateUpdateCallback(varStack)` ?

Reply:
None, i actually realised that as well, but forgot after :facepalm:. Will do